File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        res = ""
        reslength = 0
        
        for i in range(len(s)):
            
            #odd length
            l, r = i, i
            while l>=0 and r < len(s) and s[l] == s[r]:
                if (r-l+1) > reslength:
                    res = s[l:r+1]
                    reslength = r-l+1
                    
                l -= 1
                r += 1
                
            #even length
            l, r = i, i + 1
            while l>=0 and r < len(s) and s[l] == s[r]:
                if (r-l+1) > reslength:
                    res = s[l:r+1]
                    reslength = r-l+1
                    
                l -= 1
                r += 1
                
        return res

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        boolean[][] dp = new boolean[s.length()][s.length()];
        int len =0;
        String res = "";
        
        for(int gap = 0; gap < s.length(); gap++){
            for(int i = 0, j = gap; j < s.length(); i++, j++){
                if(gap == 0){
                    dp[i][j] = true;
                }
                else if(gap == 1){
                    if(s.charAt(i) == s.charAt(j)){
                        dp[i][j] = true;
                    }
                    else{
                        dp[i][j] = false;
                    }
                }
                else{
                    if(s.charAt(i)==s.charAt(j) && dp[i+1][j-1]==true){
                        dp[i][j] = true;
                    }
                    else{
                        dp[i][j] = false;
                    }
                }
                
                if(dp[i][j]){
                    len = gap + 1;
                    res = s.substring(i, j+1);
                }
            }
        }
        return res;
    }
}